File: Data Structures & Algorithms/matrixBFS/submission-0.java
class Solution {
    public int shortestPath(int[][] grid) {
        if (grid.length == 0) return -1;
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        return bfs(grid, visited);
    }

    public int bfs(int[][] grid, boolean[][] visited) {
        int rows = grid.length;
        int cols = grid[0].length;
        int length = 0;
        Deque<Integer[]> queue = new ArrayDeque<>();
        queue.add(new Integer[] {0, 0});
        visited[0][0] = true;
        int[][] moves = new int[][] {
            {1, 0}, {0, 1}, {-1, 0}, {0, -1}
        };

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {   
                Integer[] arr = queue.removeFirst();
                int r = arr[0];
                int c = arr[1];
                if (r == rows - 1 && c == cols - 1) {
                    return length;
                }
                for (int[] m : moves) {
                    int nr = r + m[0];
                    int nc = c + m[1];
                    if (Math.min(nr, nc) < 0 || nr == rows 
                    || nc == cols || grid[nr][nc] == 1 
                    || visited[nr][nc]) {
                        continue;
                    }
                    visited[nr][nc] = true;
                    queue.add(new Integer[] {nr, nc});
                }
            }
            length += 1;
        }
        return -1;
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        INF = 2147483647
        ROWS, COLS = len(grid), len(grid[0])

        def bfs(i, j):
            visited = set()
            queue = deque()
            queue.append((i, j))
            dist = 0
            while queue:
                size = len(queue)
                for _ in range(size):
                    r, c = queue.popleft()
                    if grid[r][c] == 0:
                        return dist
                    visited.add((r,c))
                    directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
                    for dr, dc in directions:
                        nr, nc = r + dr, c + dc
                        if (
                            min(nr, nc) < 0 or nr == ROWS or nc == COLS
                            or grid[nr][nc] == -1 or (nr, nc) in visited
                        ):
                            continue
                        queue.append((nr, nc))
                        visited.add((nr,nc))
                dist += 1
            return dist

        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == INF:
                    distance = bfs(row, col)
                    grid[row][col] = INF if distance == 0 else distance

